File: Day-02/typescript/zFlxw/day02.ts
import { readFileLines } from '../utils.ts';

export function day02_01() {
    const lines = readFileLines('day02/input.txt');
    let sum = 0;
    for (const line of lines) {
        if (line === '') continue;
        const [opp, mine] = line.split(' ');
        sum += getWinScore(opp, mine) + getShapeScore(mine);
    }

    console.log('1.', sum);
}

export function day02_02() {
    const lines = readFileLines('day02/input.txt');
    let sum = 0;
    for (const line of lines) {
        if (line === '') continue;
        const [opp, res] = line.split(' ');
        sum += getShapeScore(chooseShape(opp, res)) + getWinRating(res);
    }

    console.log('2.', sum);
}

function getShapeName(shape: string): string {
    switch (shape) {
        case 'A':
        case 'X':
            return 'rock';
        case 'B':
        case 'Y':
            return 'paper';
        case 'C':
        case 'Z':
            return 'scissors';
        default:
            return '';
    }
}

function getShapeScore(shape: string): number {
    const name = getShapeName(shape);
    if (name === 'rock') {
        return 1;
    } else if (name === 'paper') {
        return 2;
    } else if (name === 'scissors') {
        return 3;
    }

    return 0;
}

function getWinScore(oppShape: string, myShape: string): number {
    const myName = getShapeName(myShape);
    const oppName = getShapeName(oppShape);

    if (myName === oppName) {
        return 3;
    }

    if (
        (myName === 'rock' && oppName === 'scissors') ||
        (myName === 'paper' && oppName === 'rock') ||
        (myName === 'scissors' && oppName === 'paper')
    ) {
        return 6;
    }

    return 0;
}

function getWinRating(result: string) {
    if (result === 'Z') {
        return 6;
    }

    if (result === 'Y') {
        return 3;
    }

    return 0;
}

function chooseShape(oppShape: string, res: string) {
    const name = getShapeName(oppShape);
    if (res === 'Y') {
        return oppShape;
    }

    if (name === 'rock') {
        if (res === 'X') {
            return 'C';
        } else {
            return 'B';
        }
    } else if (name === 'paper') {
        if (res === 'X') {
            return 'A';
        } else {
            return 'C';
        }
    } else if (name === 'scissors') {
        if (res === 'X') {
            return 'B';
        } else {
            return 'A';
        }
    }

    return '';
}

File: Day-02/cpp/Hollowchick/day02.cpp
#include <iostream>
#include <fstream>
using namespace std;

int main() {
    string line;
    int score {0}, score2{0};
    ifstream input ("input");
    while (input) {
        getline(input, line);
        if (line.empty()){
            break;
        }
        switch (line[2]) {
            case 'X':
                score += 1;
                if (line[0] == 'A') { score += 3; } // Rock - Rock
                else if (line[0] == 'B') { score += 0; } // Paper - Rock
                else if (line[0] == 'C') { score += 6; } // Scissors - Rock
                break;
            case 'Y': score += 2;
                if (line[0] == 'A') { score += 6; } // Rock - Paper
                else if (line[0] == 'B') { score += 3; } // Paper - Paper
                else if (line[0] == 'C') { score += 0; } // Scissors - Paper
                break;
            case 'Z': score += 3;
                if (line[0] == 'A') { score += 0; } // Rock - Scissors
                else if (line[0] == 'B') { score += 6; } // Paper - Scissors
                else if (line[0] == 'C') { score += 3; } // Scissors - Scissors
                break;
        }
        switch (line[0]) {
            case 'A':
                if (line[2] == 'X') { score2 += 3; } // Rock - Scissors
                else if (line[2] == 'Y') { score2 += 4; } // Rock - Rock
                else if (line[2] == 'Z') { score2 += 8; } // Rock - Paper
                break;
            case 'B':
                if (line[2] == 'X') { score2 += 1; } // Paper - Rock
                else if (line[2] == 'Y') { score2 += 5; } // Paper - Paper
                else if (line[2] == 'Z') { score2 += 9; } // Paper - Scissors
                break;
            case 'C':
                if (line[2] == 'X') { score2 += 2; } // Scissors - Paper
                else if (line[2] == 'Y') { score2 += 6; } // Scissors - Scissors
                else if (line[2] == 'Z') { score2 += 7; } // Scissors - Rock
                break;
        }
    }
    cout << score << endl;
    cout << score2 << endl;
    return 0;
}
